module/extension/FileWatcher/tests/FileWatcherTestDirectory.cpp

```diff
@@ -0,0 +1,130 @@
+
+class TestReactor : public NUClear::Reactor {
+public:
+    TestReactor(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)) {
+        // GIVEN a directory containing two files.
+        std::filesystem::path dirPath = std::filesystem::temp_directory_path() / "directory-watcher-test-XXXXXX";
+        std::string dirPathString     = dirPath.string();
+
+        std::vector<char> dirPathChars(dirPathString.c_str(), dirPathString.c_str() + dirPathString.size() + 1);
+        dirPath = mkdtemp(dirPathChars.data());
+
+        std::filesystem::path file1Path = dirPath / testFile1Name;
+        std::filesystem::path file2Path = dirPath / testFile2Name;
+
+        // Create files and close streams at end of scope
+        {
+            std::ofstream(file1Path.string());
+            std::ofstream(file2Path.string());
+        }
+
+        on<extension::FileWatch>(dirPath.string(), extension::FileWatch::CHANGED)
+            .then([this, file1Path, file2Path](const extension::FileWatch& message) {
+                UNSCOPED_INFO("Directory change event: " << message.path);
+                fileWatchEvents.push_back(message);
+
+                if (fileWatchEvents.size() >= 4) { powerplant.shutdown(); }
+            });
+
+        on<Startup>().then([this] { emit(std::make_unique<StartTest>()); });
+        on<Trigger<StartTest>, MainThread>().then([this, file1Path, file2Path] {
+            // WHEN the file contents are modified.
+            std::ofstream ofs1 = std::ofstream(file1Path, std::ofstream::app);
+            ofs1 << "Test\n";
+            ofs1.close();
+            UNSCOPED_INFO("File modified: " << file1Path);
+
+            std::ofstream ofs2 = std::ofstream(file2Path, std::ofstream::app);
+            ofs2 << "Test\n";
+            ofs2.close();
+            UNSCOPED_INFO("File modified: " << file2Path);
+        });
+
+        on<Watchdog<TestReactor, 1000, std::chrono::milliseconds>>().then([this] {
+            UNSCOPED_INFO("Watchdog timeout.");
+            watchdogTriggered = true;
+            powerplant.shutdown();
+        });
+    }
+
+    virtual ~TestReactor() {
+        std::filesystem::remove_all(dirPath);  // Cleanup
+    }
+
+private:
+    std::filesystem::path dirPath;
+};
+
+bool eventTriggered(std::string fileName, extension::FileWatch::Event event) {
+    auto iterator = std::find_if(
+        fileWatchEvents.begin(), fileWatchEvents.end(), [fileName, event](extension::FileWatch const& fileWatch) {
+            std::filesystem::path filePath = fileWatch.path;
+            return filePath.filename() == fileName && fileWatch.events == event;
+        });
+
+    return iterator != fileWatchEvents.end();
+}
+
+}  // namespace
+
+
+TEST_CASE(
+    "Given a directory containing two files, when the file contents are modified, then change events are triggered for "
+    "each file.") {
+
+    INFO("Start Test");
+
+    NUClear::PowerPlant::Configuration config;
+    config.thread_count = 1;
```

Comment:
Since we know we used to have some sort of race condition is it worth adding extra threads to try and make sure it doesn't happen?

Reply:
Where the race condition was this won't make a difference